src/EFCore.PG/Storage/Internal/Mapping/NpgsqlArrayTypeMapping.cs

```diff
@@ -19,9 +23,26 @@ public abstract class NpgsqlArrayTypeMapping : RelationalTypeMapping
         /// </summary>
         public RelationalTypeMapping ElementMapping { get; }
 
+        /// <summary>
+        /// The database type used by Npgsql.
+        /// </summary>
+        public NpgsqlDbType? NpgsqlDbType { get; }
+
         protected NpgsqlArrayTypeMapping(RelationalTypeMappingParameters parameters, RelationalTypeMapping elementMapping)
             : base(parameters)
-            => ElementMapping = elementMapping;
+        {
+            ElementMapping = elementMapping;
+
+            // If the element mapping has an NpgsqlDbType or DbType, set our own NpgsqlDbType as an array of that.
+            // Otherwise let the ADO.NET layer infer the PostgreSQL type. We can't always let it infer, otherwise
+            // when given a byte[] it will infer byte (but we want smallint[])
+            NpgsqlDbType = NpgsqlTypes.NpgsqlDbType.Array |
+                           (elementMapping is NpgsqlTypeMapping elementNpgsqlTypeMapping
+                               ? elementNpgsqlTypeMapping.NpgsqlDbType
+                               : elementMapping.DbType.HasValue
+                                   ? new NpgsqlParameter { DbType = elementMapping.DbType.Value }.NpgsqlDbType
+                                   : default(NpgsqlDbType?));
```

Comment:
To large indentation, but it's a nit.

Reply:
This is how Rider likes it... I've stopped fighting :)

Reply:
As I know, it's adjustable. For me it doesn't recommend such things, am using a cleaned up editor config from `dotnet/runtime`.

Reply:
Yeah. If you really feel like it you can work on our .editorconfig :)